ydb_dbapi/cursors.py

```diff
@@ -0,0 +1,363 @@
+            if new_rows_count != -1:
+                if self._rows_count != -1:
+                    self._rows_count += new_rows_count
+                else:
+                    self._rows_count = new_rows_count
+
+    def _update_description(self, result_set: ydb.convert.ResultSet) -> None:
+        self._description = [
+            (
+                col.name,
+                _get_column_type(col.type),
+                None,
+                None,
+                None,
+                None,
+                None,
+            )
+            for col in result_set.columns
+        ]
+
+    def _raise_if_running(self) -> None:
+        if self._state == CursorStatus.running:
+            raise ProgrammingError(
+                "Some records have not been fetched. "
+                "Fetch the remaining records before executing the next query."
+            )
+
+    def _raise_if_closed(self) -> None:
+        if self.is_closed:
+            raise InterfaceError(
+                "Could not perform operation: Cursor is closed."
+            )
+
+    @property
+    def is_closed(self) -> bool:
+        return self._state == CursorStatus.closed
+
+    def _begin_query(self) -> None:
+        self._state = CursorStatus.running
+
+    def _fetchone_from_buffer(self) -> tuple | None:
+        self._raise_if_closed()
+        return next(self._rows or iter([]), None)
+
+    def _fetchmany_from_buffer(self, size: int | None = None) -> list:
+        self._raise_if_closed()
+        return list(
+            itertools.islice(self._rows or iter([]), size or self.arraysize)
+        )
+
+    def _fetchall_from_buffer(self) -> list:
+        self._raise_if_closed()
+        return list(self._rows or iter([]))
+
+
+class Cursor(BufferedCursor):
+    def __init__(
+        self,
+        session: ydb.QuerySession,
+        tx_context: ydb.QueryTxContext | None = None,
+        table_path_prefix: str = "",
+        autocommit: bool = True,
+    ) -> None:
+        self._session = session
+        self._tx_context = tx_context
+        self._table_path_prefix = table_path_prefix
+        self._autocommit = autocommit
+
+        self._stream: Iterator | None = None
+
+    def fetchone(self) -> tuple | None:
+        return self._fetchone_from_buffer()
+
+    def fetchmany(self, size: int | None = None) -> list:
+        size = size or self.arraysize
+        return self._fetchmany_from_buffer(size)
+
+    def fetchall(self) -> list:
+        return self._fetchall_from_buffer()
+
+    @handle_ydb_errors
+    def _execute_generic_query(
+        self, query: str, parameters: ParametersType | None = None
+    ) -> Iterator[ydb.convert.ResultSet]:
+        return self._session.execute(query=query, parameters=parameters)
+
+    @handle_ydb_errors
+    def _execute_transactional_query(
+        self, query: str, parameters: ParametersType | None = None
+    ) -> Iterator[ydb.convert.ResultSet]:
+        if self._tx_context is None:
```

Comment:
**suggestion**: Just move `tx_context` to the arguments and you could remove this check